fendermint/testing/contract-test/tests/upgrades.rs

```diff
@@ -0,0 +1,300 @@
+        let state = guard.take().expect("exec state empty");
+
+        let (state, ret) = f(state).await?;
+
+        *guard = Some(state);
+
+        Ok(ret)
+    }
+
+    /// Put the execution state during block execution. Has to be empty.
+    async fn put_exec_state(&self, state: FvmExecState<MemoryBlockstore>) {
+        let mut guard = self.exec_state.lock().await;
+        assert!(guard.is_none(), "exec state not empty");
+        *guard = Some(state);
+    }
+
+    /// Take the execution state during block execution. Has to be non-empty.
+    async fn take_exec_state(&self) -> FvmExecState<MemoryBlockstore> {
+        let mut guard = self.exec_state.lock().await;
+        guard.take().expect("exec state empty")
+    }
+
+    async fn begin_block(&self, block_height: ChainEpoch) -> Result<()> {
+        // generate a random block hash
+        let block_hash: [u8; 32] = [0; 32];
+
+        let db = self.state_store.as_ref().clone();
+        let mut state_params = self.state_params.clone();
+        state_params.timestamp = Timestamp(block_height as u64);
+
+        let state = FvmExecState::new(db, self.multi_engine.as_ref(), block_height, state_params)
+            .context("error creating new state")?
+            .with_block_hash(block_hash);
+
+        self.put_exec_state(state).await;
+
+        let res = self
+            .modify_exec_state(|s| self.interpreter.begin(s))
+            .await
+            .unwrap();
+        println!("Ret begin apply_ret : {:?}", res.apply_ret);
+
+        Ok(())
+    }
+
+    async fn end_block(&self, _block_height: ChainEpoch) -> Result<()> {
+        let _ret = self
+            .modify_exec_state(|s| self.interpreter.end(s))
+            .await
+            .context("end failed")?;
+
+        Ok(())
+    }
+
+    async fn commit(&mut self) -> Result<()> {
+        let exec_state = self.take_exec_state().await;
+
+        let (
+            state_root,
+            FvmUpdatableParams {
+                power_scale,
+                circ_supply,
+            },
+            _,
+        ) = exec_state.commit().context("failed to commit FVM")?;
+
+        self.state_params.state_root = state_root;
+        self.state_params.power_scale = power_scale;
+        self.state_params.circ_supply = circ_supply;
+
+        Ok(())
+    }
+}
+
+#[test]
+fn testest() {
+    let rt: tokio::runtime::Runtime =
+        tokio::runtime::Runtime::new().expect("create tokio runtime for init");
+
+    let mut us = UpgradeScheduler::default();
+
+    let chain_id = 1942764459484029;
+
+    us.add(Upgrade {
+        chain_id,
+        block_height: 1,
+        migration: |state| {
+            println!(
+                "Migration at height {} just prints out chainmetadata actor",
+                state.block_height()
+            );
+
+            let actor = state
+                .state_tree_mut()
+                .get_actor(system::SYSTEM_ACTOR_ID)
+                .unwrap();
```

Comment:
```suggestion
                .get_actor(system::SYSTEM_ACTOR_ID)?;
```

It looks like this closure should be able to handle errors. I think this test might be one that any prospective upgrade author would look at, so we should try to be as close as possible to how we'd like migrations to be written.

It will result in a panic but at least we'd know where it happens: in the ABCI service wrapper.